### PR TITLE
PoC of in-place node upgrades

### PR DESCRIPTION
**Pull request by jrryjcksn**
Orchestrates simple upgrade by:
* Upgrading the initial master in place
* Upgrading the other masters
* Upgrading workers

The ordering is supported via the same sort of approach in the existing controller (i.e. skipping an update if there is another node with higher priority -- the priorities being defined as above).

This PR serves as a proof of concept.

### CODE REVIEW

**Comment by fbarl** [comment]
> There is much to be done before this could be considered anywhere near product quality.

@jrryjcksn I'd be curious if you could write down what's currently missing to get us there, i.e. try to list all the bigger chunks of work. I imagine that information might be useful for the upcoming sprint plannings :)

File: pkg/apis/wksprovider/controller/wksctl/cluster_actuator.go [hunk 1]
```diff
@@ -21,7 +21,7 @@ type ClusterActuator struct {
 // Reconcile creates or updates the cluster.
 func (a *ClusterActuator) Reconcile(cluster *clusterv1.Cluster) error {
 	a.recordEvent(cluster, corev1.EventTypeNormal, "Reconcile", "Reconciled cluster %v", cluster.Name)
-	return errors.New("ClusterActuator#Reconcile not implemented")
+	return nil
```
Comment (chanwit) [inline pkg/apis/wksprovider/controller/wksctl/cluster_actuator.go]: this is fine to me as it's a block that I'm also facing.

**Comment by kingdonb** [comment]
I am tracking this in #89,

In order to get my ImageTag to build, I have set the mtime of the resources in `pkg/apis/cluster-api/config/crds/`

You can do this and `go generate` will create a matching `pkg/apis/wksprovider/machine/crds/crds_vfsdata.go` in order to get that diff to go away, I have done this so that a binary will be built that does not include a -WIP string in controller's ImageTag setting.

```
touch -t '201908141313.26' pkg/apis/cluster-api/config/crds/ pkg/apis/cluster-api/config/crds/*
```

I have used this version to successfully bring up a multimaster cluster, it reached a steady state, but I am speculating that I will still need to take steps to enable a stable controlPlaneEndpoint, because as soon as I kick off the upgrade for the first time, I find the control plane is not responding anymore.  It comes back after some time, then drops out again, but never seems to progress past the first (node0) upgrade.

Last time, I was able to use my own fork of footloose and a loadbalancer docker container to make a stable controlPlaneEndpoint for the cluster; I don't know if this still is necessary, but replaying that plan is my next step, since I don't see any solution for that issue in this PR.

How are you testing this works, for my curiosity? (I guess you aren't using the Footloose mode?  Does the GCP version provision a loadbalancer and use it as controlPlaneEndpoint?)

**Comment by kingdonb** [comment]
I've been advised this should work with footloose, but when I tested it I had trouble twice, first time most likely due to memory as it tanked before all nodes even came online (this was with the default memory on Docker for Mac).  There might be some minimal additional documentation about system reqs needed to ensure that newcomers can succeed at upgrading on the first try.

My understanding is, though, it should work even without a stable controlPlaneEndpoint, the masters should all be able to find each other even after the leading master has gone down for upgrade, and the upgrade should succeed regardless of there being no control plane LB.

This morning I've tried an upgrade to v1.14.10 with the footloose multimaster example and used a docker daemon with about 10GB of RAM allocated, made sure I have the latest version of footloose from git, wksctl built from this branch with the changes to `crds_vfsdata.go` mtime carefully zeroed out, and here's my report (it's good news):

1. The masters reach a stable state and the worker joins, everything OK, wks-controller reports "nothing to do"
2. `kubectl edit machines` to locate the versions section of the spec, I find "kubelet" version is set to an empty string on all four nodes. I replace it with a `kubelet: "1.14.1"` for the worker, and a 
```
kubelet: "1.14.1"
controlPlane: "1.14.1"
```
for each master, which has the desired effect of again, wks-controller reports nothing to do.

3. Finally I edit machines again and bump the versions which I just pinned, now setting them all at once to 1.14.10. The wks-controller is immediately evicted and placed onto another master, and less than 10 minutes later I have seen it moving smoothly onto each successive master before it gets upgraded in turn, now the cluster appears to be upgraded! Awesome.

```
NAME    STATUS   ROLES    AGE     VERSION
node0   Ready    master   18m     v1.14.10
node1   Ready    master   14m     v1.14.10
node2   Ready    master   11m     v1.14.10
node3   Ready    <none>   8m46s   v1.14.10
```

For good measure I repeat the experiment with an upgrade to 1.15.8, at no point do I see the control plane dropping out for an extended period, and I'm not sure how that happens without a load balancer, it almost looks like you've actually routed the kubectl client traffic through the worker? It's not clear to me that I've understood this correctly, here's what I'm reading to get that interpretation:

`kubeconfig` excerpt:
```
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: LS0tLS...=
    server: https://127.0.0.1:6443
```

```
$ footloose show -c centos7/docker/multimaster.yaml
NAME                       HOSTNAME   PORTS                         IP           IMAGE                             CMD          STATE     BACKEND
centos-multimaster-node0   node0      0->{22 2222},1->{6443 6443}   172.17.0.2   quay.io/footloose/centos7:0.3.0   /sbin/init   Running
centos-multimaster-node1   node1      0->{22 2223},1->{6443 6444}   172.17.0.3   quay.io/footloose/centos7:0.3.0   /sbin/init   Running
centos-multimaster-node2   node2      0->{22 2224},1->{6443 6445}   172.17.0.4   quay.io/footloose/centos7:0.3.0   /sbin/init   Running
centos-multimaster-node3   node3      0->{22 2225},1->{6443 6446}   172.17.0.5   quay.io/footloose/centos7:0.3.0   /sbin/init   Running
```

I don't quite understand how this works yet but I can verify that it does!  Except for some strange reason, the second upgrade attempt to `1.15.8` has left me with a failed state, one master has upgraded to `node0   Ready    master   30m   v1.15.9-beta.0` and the upgrade stops in its tracks indicating failure.  (Which is another kind of success in its own way, I guess)

**Comment by kingdonb** [comment]
Looks like I should have set some versions on the cluster.yml as well...

**Comment by kingdonb** [comment]
It looks like I jumped the gun on 1.15.9, my trouble was because it wasn't really released yet at the time of my last report / only actually came out 5 days ago.

I took the new alpha build over to wks-firekube quickstart and tried some version upgrading there, had better results. I was able to upgrade a running cluster from 1.14.1 to 1.14.10, then to 1.15.9, and next attempted upgrading onward to 1.16.6, all succeeded until the last (1.16.6 failed) on a footloose cluster with 3 control plane and 1 worker node.

Only minimal changes to the firekube were needed to get this far (I opened weaveworks/wks-quickstart-firekube#75).  I only needed to make changes to machine specs to perform upgrades, setting spec.versions.kubelet and spec.versions.controlPlane in setup.js. (not needed to change cluster.yaml as I had speculated, I also guessed that "controlPlane" was needed, but not sure.)

It would probably be much less resource intensive to just stand the cluster up with the desired version from the outset, (and I'll try that next) – especially as these particular clusters are meant to be disposable, but it's good to know that firekube can perform upgrades on a running cluster without balking or crashing!

--- 

(notes regarding upgrade to 1.16.6 that failed, this is probably an expected failure)

I may have made a mistake in my manifests and accidentally skipped one node for upgrade, but I think it wouldn't have mattered.

Other than a brief period of time when I couldn't tail `deploy/wks-controller` logs because the API was in mid-upgrade of those deprecated API groups, everything went totally smoothly with the upgrade, and those things resolved themselves by the time my first master was upgraded to 1.16.6. The 1.16.6 control plane seemed to be functioning fine, the trouble came when the next master tried to upgrade, master-2, seems to have failed on or around `kubeadm join`:

```
time="2020-01-26T18:42:50Z" level=info msg="........................NEW UPDATE FOR: master-2..........................." cluster=example machine=master-2 node=node2
time="2020-01-26T18:42:50Z" level=debug msg="Kubernetes version used" machine=master-2 version=1.16.6
time="2020-01-26T18:42:50Z" level=info msg="Is master: true, is up or downgrade: true" cluster=example machine=master-2 node=node2
time="2020-01-26T18:42:50Z" level=debug msg="Kubernetes version used" machine=master-2 version=1.16.6
time="2020-01-26T18:42:50Z" level=debug msg="Kubernetes version used" machine=master-2 version=1.16.6
time="2020-01-26T18:42:50Z" level=info msg="Original needs update: false" cluster=example machine=master-2 node=node2
time="2020-01-26T18:42:51Z" level=info msg="Master needs update: true" cluster=example machine=master-2 node=node2
time="2020-01-26T18:42:51Z" level=info msg="Is original: false" cluster=example machine=master-2 node=node2
time="2020-01-26T18:42:51Z" level=info msg="Is controller: false" cluster=example machine=master-2 node=node2
time="2020-01-26T18:42:51Z" level=info msg=Undoing resource="upgrade:node-lock-kubernetes"
time="2020-01-26T18:42:51Z" level=info msg=Undone resource="upgrade:node-lock-kubernetes"
time="2020-01-26T18:42:51Z" level=info msg=Undoing resource="upgrade:node-kubectl"
time="2020-01-26T18:42:51Z" level=info msg=Undone resource="upgrade:node-kubectl"
time="2020-01-26T18:42:51Z" level=info msg=Undoing resource="upgrade:node-restart-kubelet"
time="2020-01-26T18:42:51Z" level=info msg=Undone resource="upgrade:node-restart-kubelet"
time="2020-01-26T18:42:51Z" level=info msg=Undoing resource="upgrade:node-kubelet"
time="2020-01-26T18:42:51Z" level=info msg=Undone resource="upgrade:node-kubelet"
time="2020-01-26T18:42:51Z" level=info msg=Undoing resource="upgrade:node-kubeadm-upgrade"
time="2020-01-26T18:42:51Z" level=info msg=Undone resource="upgrade:node-kubeadm-upgrade"
time="2020-01-26T18:42:51Z" level=info msg=Undoing resource="upgrade:node-install-kubeadm"
time="2020-01-26T18:42:51Z" level=info msg=Undone resource="upgrade:node-install-kubeadm"
time="2020-01-26T18:42:51Z" level=info msg=Undoing resource="upgrade:node-unlock-kubernetes"
time="2020-01-26T18:42:51Z" level=info msg=Undone resource="upgrade:node-unlock-kubernetes"
time="2020-01-26T18:42:51Z" level=info msg=Starting resource="upgrade:node-unlock-kubernetes"
time="2020-01-26T18:42:51Z" level=debug msg="running command: sudo -n -- sh -c 'yum versionlock delete '\"'\"'kube*'\"'\"' || true'"
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/binHOSTNAME=node2TERM=xtermcontainer=dockerHOME=/rootLoaded plugins: fastestmirror, ovl, versionlock
Error: Error: versionlock delete: no matches
time="2020-01-26T18:42:52Z" level=info msg=Finishing resource="upgrade:node-unlock-kubernetes"
time="2020-01-26T18:42:52Z" level=info msg=Starting resource="upgrade:node-install-kubeadm"
time="2020-01-26T18:42:52Z" level=debug msg="running command: sudo -n -- sh -c 'rpm -q --queryformat '\"'\"'%{NAME} %{VERSION} %{RELEASE}\\n'\"'\"' kubeadm-1.16.6'"
kubeadm 1.16.6 0
time="2020-01-26T18:42:52Z" level=info msg=Finishing resource="upgrade:node-install-kubeadm"
time="2020-01-26T18:42:52Z" level=info msg=Starting resource="upgrade:node-kubeadm-upgrade"
time="2020-01-26T18:42:52Z" level=debug msg="running command: sudo -n -- sh -c 'kubeadm upgrade node experimental-control-plane'"
Error: unknown command "experimental-control-plane" for "kubeadm upgrade node"
Usage:
  kubeadm upgrade node [flags]
  kubeadm upgrade node [command]

Available Commands:
...
```

Kubeadm seems to be the right version (v1.16.6) on both upgraded master nodes.

I think there is some special handling needed to get past this upgrade. Not clear to me what blocks this, but I have a guess after reading the above... I bet kubeadm docs can answer this clearer, but `kubeadm join --experimental-control-plane` looks like it was the way to say it in v1.14, and v1.15 perhaps still permitted this, but as of v1.16 it has probably just become `--control-plane`.

I realize the actual operation that failed is `kubeadm upgrade node experimental-control-plane`, probably just about the same thing in principle.

Maybe I can take another look later and give the PR a try... I don't know much about wksctl internals yet, but the place to start looks like `pkg/plan/resource/kubeadm_join.go`

Regardless of any troubles with later versions, the Proof of Concept is proven IMHO, this change definitely works for me in 0.8.2-alpha3 👍

**Comment by kingdonb** [comment]
For sure, there are more issues to tackle before even 1.15.9 can be installed out of the box, from the looks of things. I'm not sure why I was able to upgrade, but can't install from scratch at 1.15.9. It's going to take more than just removing the protecting validation that stopped me from trying at first.

The issues seem to mostly revolve around Weave net and CNI failing to come up, which I tried upgrading in the addons to a version more appropriate for 1.15.9, but gave up when I saw errors coming up in weave-net related to netfilter and insmod:

cat: can't open '/proc/sys/net/bridge/bridge-nf-call-iptables': No such file or directory

I haven't been through everything yet, might be some other things I have missed during the upgrade.

**Pull request closed without merging**